crates/rust-analyzer/src/reload.rs

```diff
@@ -234,19 +247,40 @@ impl GlobalState {
                     .iter()
                     .map(|project| match project {
                         LinkedProject::ProjectManifest(manifest) => {
+                            debug!(path = %manifest, "loading project from manifest");
+
                             project_model::ProjectWorkspace::load(
                                 manifest.clone(),
                                 &cargo_config,
                                 &progress,
                             )
                         }
                         LinkedProject::InlineJsonProject(it) => {
-                            Ok(project_model::ProjectWorkspace::load_inline(
+                            let (Some(_command), Some(path)) = (&discover_command, &path) else {
+                                let workspace = project_model::ProjectWorkspace::load_inline(
+                                    it.clone(),
+                                    cargo_config.target.as_deref(),
+                                    &cargo_config.extra_env,
+                                    &cargo_config.cfg_overrides,
+                                );
+                                #[allow(clippy::needless_return)]
+                                return Ok(workspace);
+                            };
+
+                            if let Some(build) = it.crate_by_buildfile(path) {
+                                if is_quiescent {
+                                    let arg = DiscoverProjectParam::Label(build.label);
+                                    sender.send(Task::DiscoverLinkedProjects(arg)).unwrap();
+                                }
+                            }
```

Comment:
clippy is unhappy about this (as am I) but I _think_ fixing this requires moving the `discover_command` functionality to live in `ProjectWorkspace`, alongside some additional changes in `ProjectWorkspace` to avoid repeatedly calling `discover_command` and rely on some internally-cached `rust-project.json` when `GlobalState::fetch_workspaces` is called.

Reply:
this is gross, but it's a bit less gross now.